libs/cli/src/main/java/org/elasticsearch/cli/internal/JsonPrintWriter.java

```diff
@@ -0,0 +1,214 @@
+
+    @SuppressForbidden(reason = "Override PrintWriter to emit Json")
+    protected JsonPrintWriter(Map<String, String> staticFields, OutputStream out, boolean autoFlush, Clock clock) {
+        super(out, autoFlush);
+        this.clock = clock;
+        this.buffer = new StringBuilder(BUFFER_INITIAL_SIZE);
+        this.bufferPreambleSize = initStaticPreamble(buffer, staticFields);
+    }
+
+    private static int initStaticPreamble(StringBuilder builder, Map<String, String> fields) {
+        builder.append('{');
+        fields.forEach((name, value) -> builder.append('"').append(name).append("\":\"").append(value).append("\","));
+        builder.append("\"@timestamp\":\"");
+        return builder.length();
+    }
+
+    private void initMessageIfAtPreamble() {
+        assert Thread.holdsLock(buffer); // only allow if holding lock on buffer
+        assert buffer.length() >= bufferPreambleSize;
+
+        if (buffer.length() == bufferPreambleSize) {
+            ISO_INSTANT_FORMATTER.formatTo(clock.instant(), buffer);
+            buffer.append("\", \"message\":\"");
+        }
+    }
+
+    @Override
+    public void write(int c) {
+        synchronized (buffer) {
+            initMessageIfAtPreamble();
+            JsonUtils.quote((char) c, buffer);
+        }
+    }
+
+    @Override
+    public void write(char[] chars, int off, int len) {
+        synchronized (buffer) {
+            initMessageIfAtPreamble();
+            JsonUtils.quoteAsString(chars, off, len, buffer);
+        }
+    }
+
+    @Override
+    public void write(String msg, int off, int len) {
+        synchronized (buffer) {
+            initMessageIfAtPreamble();
+            JsonUtils.quoteAsString(msg, off, len, buffer);
+        }
+    }
+
+    @Override
+    public void println(String msg) {
+        synchronized (lock) {
+            if (singleLineMode == false && isJsonObject(msg)) {
+                super.write(msg, 0, msg.length());
+                super.println();
+            } else {
+                synchronized (buffer) {
+                    write(msg, 0, msg.length());
+                    println();
+                }
+            }
+        }
+    }
+
+    @Override
+    public void println(char[] chars) {
+        synchronized (lock) {
+            if (singleLineMode == false && isJsonObject(chars)) {
+                // Forward as is, ignore current buffer.
+                super.write(chars, 0, chars.length);
+                super.println();
+            } else {
+                synchronized (buffer) {
+                    write(chars, 0, chars.length);
+                    println();
+                }
+            }
+        }
+    }
+
+    /** Print an exception (with stacktrace) into a single JSON object. */
+    public void println(Exception exception) {
+        synchronized (lock) {
+            synchronized (buffer) {
+                singleLineMode = true;
+                try {
+                    exception.printStackTrace(this);
```

Comment:
Could we do better here, using the ecs error fields like error.message and error.stack_trace?

Reply:
Done 👍